scripts/generate-clients.mjs

```diff
@@ -7,66 +7,95 @@ import { renderVisitor as renderRustVisitor } from "@kinobi-so/renderers-rust";
 import { getAllProgramIdls } from "./utils.mjs";
 
 // Instanciate Kinobi.
-const [idl, ...additionalIdls] = getAllProgramIdls().map(idl => rootNodeFromAnchor(require(idl)))
+const [idl, ...additionalIdls] = getAllProgramIdls().map((idl) =>
+  rootNodeFromAnchor(require(idl))
+);
 const kinobi = k.createFromRoot(idl, additionalIdls);
 
 // Update programs.
 kinobi.update(
   k.updateProgramsVisitor({
-    "stakeProgram": { name: "stake" },
+    stakeProgram: { name: "stake" },
   })
 );
 
-// Update accounts.
+// Add PDA information.
 kinobi.update(
-  k.updateAccountsVisitor({
-    counter: {
-      seeds: [
-        k.constantPdaSeedNodeFromString("utf8", "counter"),
-        k.variablePdaSeedNode(
-          "authority",
-          k.publicKeyTypeNode(),
-          "The authority of the counter account"
-        ),
-      ],
+  k.bottomUpTransformerVisitor([
+    {
+      select: "[programNode]stake",
+      transform: (node) => {
+        k.assertIsNode(node, "programNode");
+        return {
+          ...node,
+          pdas: [
+            k.pdaNode({
+              name: "vault",
+              seeds: [
+                k.constantPdaSeedNodeFromString("utf8", "token-vault"),
+                k.variablePdaSeedNode(
+                  "authority",
+                  k.publicKeyTypeNode(),
+                  "Config authority"
+                ),
+              ],
+            }),
+          ],
+        };
+      },
     },
-  })
+  ])
 );
 
-// Update instructions.
+// Add missing types from the IDL.
 kinobi.update(
-  k.updateInstructionsVisitor({
-    create: {
-      byteDeltas: [k.instructionByteDeltaNode(k.accountLinkNode("counter"))],
-      accounts: {
-        counter: { defaultValue: k.pdaValueNode("counter") },
-        payer: { defaultValue: k.accountValueNode("authority") },
+  k.bottomUpTransformerVisitor([
+    {
+      // OptionalNonZeroPubkey -> NullableAddress
+      select: (node) => {
+        const names = ["authority", "slashAuthority"];
+        return (
+          names.includes(node.name) &&
+          k.isNode(node, ["structFieldTypeNode"]) &&
+          k.isNode(node.type, "definedTypeLinkNode") &&
+          node.type.name === "optionalNonZeroPubkey"
+        );
+      },
+      transform: (node) => {
+        k.assertIsNode(node, "structFieldTypeNode");
+        return {
+          ...node,
+          type: k.definedTypeLinkNode("nullableAddress", "hooked"),
+        };
```

Comment:
Same here on the Kinobi context. I'm not really sure what the `isNode` are doing in the conditional. The rest appears like it's telling Kinobi to map `OptionalNonZeroPubkey` to the nullable address type/codec you've defined in the hooks?

Reply:
Yep, exactly. In this case the selector is a bit more complex but its purpose is to select nodes with the specified `"authority"` or `"slashAuthority"`; they need to be a `structFieldTypeNode` (so even if the name of the node matches a node that it not a field of a struct, it won't be selected); they need to be a "defined type" node, which means they need to be representing a type and the type name must match `"optionalNonZeroPubkey"`.

The `bottomUpTransformerVisitor` traverses the tree and you can specify a `select` expression to select nodes and apply a transformation.

In this case the transformation is to map the nodes with type `optionalNonZeroPubkey` to the `nullableAddress` that we are providing.

Reply:
PS: the brackets on the `["structFieldTypeNode"]` is not necessary since we are only specifying a single value – I will remove them.

Reply:
Makes sense, thanks!